typed-th-in-ghc-9/article.md

```diff
@@ -0,0 +1,205 @@
+      In a case alternative: "DEVELOPMENT" -> [|| Development ||]
+   |
+20 |       "DEVELOPMENT" -> [|| Development ||]
+   |                        ^^^^^^^^^^^^^^^^^^^
+
+TH.hs:21:23: error:
+    • Couldn't match type ‘Code m2’ with ‘Q’
+      Expected: Q (TExp LogLevel)
+        Actual: Code m2 LogLevel
+    • In the Template Haskell quotation [|| Production ||]
+      In the expression: [|| Production ||]
+      In a case alternative: "PRODUCTION" -> [|| Production ||]
+   |
+21 |       "PRODUCTION" -> [|| Production ||]
+   |                       ^^^^^^^^^^^^^^^^^^
+Failed, no modules loaded.
+```
+
+In the next sections, we will see how to port this code to GHC 9 without and with backward compatibility with GHC 8.
+
+## Porting old TTH code without backward compatibility
+
+For this section, you need a GHC whose version is at least 9. Make sure to import the appropriate definitions from `template-haskell` as well:
+
+```hs
+import Language.Haskell.TH.Syntax (Code, Q, examineCode, liftCode)
+```
+
+Since we now need to return a value of type `Code`, we change our type signature. And now we can fix the code simply by using two new functions, namely `examineCode` and `liftCode`:
+
+```diff
+-logLevelFromFlag :: Q (TExp LogLevel)
+-logLevelFromFlag = do
++logLevelFromFlag :: Code Q LogLevel
++logLevelFromFlag = liftCode $ do  -- liftCode will turn 'Q' into 'Code'
+  flag <- liftIO $ lookupEnv "LOG_LEVEL"
+  case flag of
++    -- examineCode will turn 'Code' into 'Q'
+-    Nothing -> [|| Production ||]  -- We default to Production if the flag is unset
++    Nothing -> examineCode [|| Production ||]  -- We default to Production if the flag is unset
+    Just level -> case level of
+_      "DEVELOPMENT" -> [|| Development ||]
+_      "PRODUCTION" -> [|| Production ||]
++      "DEVELOPMENT" -> examineCode [|| Development ||]
++      "PRODUCTION" -> examineCode [|| Production ||]
+      other -> fail $ "Unrecognized LOG_LEVEL flag: " <> other
+```
+
+The errors in the previous section happened because the type returned by the quotations is no longer a `Q`, but `Code` instead. The solution is to use `examineCode` to turn a `Code m a` into a `m (TExp a)`, and then `liftCode` to do the opposite operation: turning a `m (TExp a)` into a `Code m a`.
+
+```hs
+liftCode    :: forall (r :: RuntimeRep) (a :: TYPE r) m. m (TExp a) -> Code m a
+examineCode :: forall (r :: RuntimeRep) (a :: TYPE r) m. Code m a -> m (TExp a)
+```
+
+And that's it! If you want the code to still work on GHC 8, however, make sure to read the next section.
+
+## Porting old TTH code with backward compatibility
+
+For this section, you may choose to use either GHC 8 or GHC 9. We will use the [`th-compat`](https://hackage.haskell.org/package/th-compat) package, besides the familiar [`template-haskell`](https://hackage.haskell.org/package/template-haskell) package, so make sure you load them.
+
+The pattern is pretty similar to the workflow with `Code`, albeit with `Splice` now.
```

Comment:
Kinda feels like some intro is missing here. Like, just a couple of words about this package providing `Splice`, which is magically defined in such a way that it works on either version of GHC.